Conflicts with existing generated files

Using the same target directory for code generation for different specifications (or variants of a specification without clean up between two consecutive runs) can lead to conflicts, which just get visible during compilation. We should consider showing a warning or even abort when there are already files in the target directory.

I just noticed the changes introduced with #1197 make `make generate` fail as it tries to generate multiple different packages into the same directory but then fails because common files like `rflx-rflx_arithmetic.ads` already exist:
```
make generate                     
tools/generate_spark_test_code.py
Parsing tests/data/specs/ethernet.rflx
Parsing tests/data/specs/in_ethernet.rflx
Parsing tests/data/specs/ipv4.rflx
Parsing tests/data/specs/in_ipv4.rflx
Parsing tests/data/specs/udp.rflx
Parsing tests/data/specs/icmp.rflx
Parsing tests/data/specs/ipv4.rflx
Parsing tests/data/specs/udp.rflx
Processing ICMP
Processing UDP
Processing IPv4
Processing Ethernet
Processing In_IPv4
Processing In_Ethernet
Generating ICMP::Tag
...
Creating tests/spark/generated/rflx-rflx_arithmetic.adb
Creating tests/spark/generated/rflx-rflx_generic_types-generic_operations.adb
Creating tests/spark/generated/rflx-rflx_message_sequence.adb
Creating tests/spark/generated/rflx-rflx_scalar_sequence.adb
Creating tests/spark/generated/rflx.ads
Generating TLV::Tag
Generating TLV::Length
Generating Derivation::Message
Creating tests/spark/generated/rflx-tlv.ads
Creating tests/spark/generated/rflx-derivation.ads
Creating tests/spark/generated/rflx-derivation-message.ads
Creating tests/spark/generated/rflx-derivation-message.adb
Creating tests/spark/generated/rflx-rflx_arithmetic.ads
Traceback (most recent call last):
  File "/.../RecordFlux/tools/generate_spark_test_code.py", line 75, in <module>
    sys.exit(main())
  File "/.../RecordFlux/tools/generate_spark_test_code.py", line 43, in main
    Generator(
  File "/.../RecordFlux/rflx/generator/generator.py", line 143, in generate
    self._write_files(units, directory, library_files, top_level_package)
  File "/.../RecordFlux/rflx/generator/generator.py", line 154, in _write_files
    self._write_library_files(directory)
  File "/.../RecordFlux/rflx/generator/generator.py", line 166, in _write_library_files
    create_file(
  File "/.../RecordFlux/rflx/generator/generator.py", line 1471, in create_file
    fail(f"file {filename} already exists", subsystem=Subsystem.GENERATOR)
  File "/.../RecordFlux/rflx/error.py", line 189, in fail
    _fail(RecordFluxError(), message, subsystem, severity, location)
  File "/.../RecordFlux/rflx/error.py", line 209, in _fail
    error.propagate()
  File "/.../RecordFlux/rflx/error.py", line 162, in propagate
    raise self
rflx.error.RecordFluxError: generator: error: file tests/spark/generated/rflx-rflx_arithmetic.ads already exists
make: *** [Makefile:162: generate] Error 1
```

Also, some tests are failing on `main` (and thus all rebased PRs) now:

```console
=========================== short test summary info ============================
FAILED tests/integration/specification_model_generator_test.py::test_definite_message_with_builtin_type
FAILED tests/integration/specification_model_generator_test.py::test_definite_parameterized_message
FAILED tests/integration/specification_model_generator_test.py::test_parameterized_message
============= 3 failed, 23 passed, 8 skipped in 229.52s (0:03:49) ==============
```

This wasn't detected by the CI, because the verification tests were skipped.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -99,7 +99,7 @@ jobs:
   installation:
     name: Installation
     needs: skip_check_general
-    if: ${{ needs.skip_check_python.outputs.should_skip != 'true' }}
+    if: ${{ needs.skip_check_general.outputs.should_skip != 'true' }}
     runs-on: ubuntu-20.04
     steps:
     - uses: actions/checkout@v2
@@ -464,7 +464,7 @@ jobs:
       - id: skip_check
         uses: fkirc/skip-duplicate-actions@v3.4.0
         with:
-          paths: '[".github/workflows/**", "*.gpr", "tests/spark/**"]'
+          paths: '[".github/workflows/**", "*.gpr", "rflx/**", "tests/**"]'
 
   verification_spark:
     name: Verification

--- a/tests/integration/specification_model_generator_test.py
+++ b/tests/integration/specification_model_generator_test.py
@@ -446,9 +446,7 @@ def test_refinement_with_self(tmp_path: Path) -> None:
     )
 
 
-@pytest.mark.verification
-def test_definite_message_with_builtin_type(tmp_path: Path) -> None:
-    spec = """\
+DEFINITE_MESSAGE_WITH_BUILTIN_TYPE_SPEC = """\
        package Test is
 
           type Length is range 0 .. 2 ** 7 - 1 with Size => 7;
@@ -464,8 +462,17 @@ def test_definite_message_with_builtin_type(tmp_path: Path) -> None:
 
        end Test;
         """
-    utils.assert_compilable_code_string(spec, tmp_path)
-    utils.assert_provable_code_string(spec, tmp_path, units=["rflx-test-message"])
+
+
+def test_definite_message_with_builtin_type(tmp_path: Path) -> None:
+    utils.assert_compilable_code_string(DEFINITE_MESSAGE_WITH_BUILTIN_TYPE_SPEC, tmp_path)
+
+
+@pytest.mark.verification
+def test_definite_message_with_builtin_type_provability(tmp_path: Path) -> None:
+    utils.assert_provable_code_string(
+        DEFINITE_MESSAGE_WITH_BUILTIN_TYPE_SPEC, tmp_path, units=["rflx-test-message"]
+    )
 
 
 def test_message_expression_value_outside_type_range(tmp_path: Path) -> None:
@@ -538,9 +545,7 @@ def test_feature_integration(tmp_path: Path) -> None:
     utils.assert_compilable_code_specs([SPEC_DIR / "feature_integration.rflx"], tmp_path)
 
 
-@pytest.mark.verification
-def test_parameterized_message(tmp_path: Path) -> None:
-    spec = """\
+PARAMETERIZED_MESSAGE_SPEC = """\
         package Test is
 
            type Length is range 1 .. 2 ** 14 - 1 with Size => 16;
@@ -559,13 +564,20 @@ def test_parameterized_message(tmp_path: Path) -> None:
 
         end Test;
         """
-    utils.assert_compilable_code_string(spec, tmp_path)
-    utils.assert_provable_code_string(spec, tmp_path, units=["rflx-test-message"])
+
+
+def test_parameterized_message(tmp_path: Path) -> None:
+    utils.assert_compilable_code_string(PARAMETERIZED_MESSAGE_SPEC, tmp_path)
 
 
 @pytest.mark.verification
-def test_definite_parameterized_message(tmp_path: Path) -> None:
-    spec = """\
+def test_parameterized_message_provability(tmp_path: Path) -> None:
+    utils.assert_provable_code_string(
+        PARAMETERIZED_MESSAGE_SPEC, tmp_path, units=["rflx-test-message"]
+    )
+
+
+DEFINITE_PARAMETERIZED_MESSAGE_SPEC = """\
         package Test is
 
            type Length is range 1 .. 2 ** 14 - 1 with Size => 16;
@@ -578,8 +590,17 @@ def test_definite_parameterized_message(tmp_path: Path) -> None:
 
         end Test;
         """
-    utils.assert_compilable_code_string(spec, tmp_path)
-    utils.assert_provable_code_string(spec, tmp_path, units=["rflx-test-message"])
+
+
+def test_definite_parameterized_message(tmp_path: Path) -> None:
+    utils.assert_compilable_code_string(DEFINITE_PARAMETERIZED_MESSAGE_SPEC, tmp_path)
+
+
+@pytest.mark.verification
+def test_definite_parameterized_message_provability(tmp_path: Path) -> None:
+    utils.assert_provable_code_string(
+        DEFINITE_PARAMETERIZED_MESSAGE_SPEC, tmp_path, units=["rflx-test-message"]
+    )
 
 
 def test_session_type_conversion_in_assignment(tmp_path: Path) -> None:

--- a/tools/generate_spark_test_code.py
+++ b/tools/generate_spark_test_code.py
@@ -2,13 +2,16 @@
 
 """Generate the SPARK code for the test project and all feature tests."""
 
+from __future__ import annotations
+
 import filecmp
 import logging
-import sys
 from pathlib import Path
 
+from rflx.common import unique
 from rflx.generator import Generator
 from rflx.integration import Integration
+from rflx.model import Model, Session, Type
 from rflx.specification import Parser
 from tests.const import SPEC_DIR
 from tests.unit.generator_test import MODELS
@@ -35,17 +38,25 @@ FEATURE_TESTS = [
 ]
 
 
-def main() -> int:
+def main() -> None:
+    generate_spark_tests()
+    generate_feature_tests()
+
+
+def generate_spark_tests() -> None:
+    remove_ada_files(OUTPUT_DIRECTORY)
+
     parser = Parser()
     parser.parse(*SPECIFICATION_FILES)
+    model = merge_models([parser.create_model(), *MODELS])
+    Generator(
+        "RFLX",
+        reproducible=True,
+        ignore_unsupported_checksum=True,
+    ).generate(model, Integration(), OUTPUT_DIRECTORY)
 
-    for model in [parser.create_model(), *MODELS]:
-        Generator(
-            "RFLX",
-            reproducible=True,
-            ignore_unsupported_checksum=True,
-        ).generate(model, Integration(), OUTPUT_DIRECTORY)
 
+def generate_feature_tests() -> None:
     generate(SHARED_DIRECTORY)
     shared_files = {f.name: f for f in (SHARED_DIRECTORY / "generated").iterdir()}
 
@@ -56,12 +67,13 @@ def main() -> int:
                 f.unlink()
                 f.symlink_to(f"../../shared/generated/{f.name}")
 
-    return 0
-
 
 def generate(feature_test: Path) -> None:
     output_directory = feature_test / "generated"
     output_directory.mkdir(exist_ok=True)
+
+    remove_ada_files(output_directory)
+
     parser = Parser()
     parser.parse(feature_test / "test.rflx")
     Generator(
@@ -71,5 +83,22 @@ def generate(feature_test: Path) -> None:
     ).generate(parser.create_model(), parser.get_integration(), output_directory)
 
 
+def remove_ada_files(directory: Path) -> None:
+    for f in directory.glob("*.ad?"):
+        print(f"Removing {f}")
+        f.unlink()
+
+
+def merge_models(models: list[Model]) -> Model:
+    types: list[Type] = []
+    sessions: list[Session] = []
+
+    for m in models:
+        types.extend(m.types)
+        sessions.extend(m.sessions)
+
+    return Model(list(unique(types)), list(unique(sessions)))
+
+
 if __name__ == "__main__":
-    sys.exit(main())
+    main()

